### DISCUSSION TITLE
Moving shapes within a sketch after creation

### DISCUSSION THREAD

**Discussion by weerasuriya** [gumyr/build123d#744] [General]
Is it possible to create a shape in a sketch (e.g., the default Plane.XY) and then move it?

For example, why do these two examples produce separate results:

```python
# Example 1
# %%
r1 = Rectangle(1,1)
r2 = Rectangle(2,2)
r2.locate(Pos(5,5))

with BuildSketch() as sk:
    add(r1)
    add(r2)

show(sk.sketch)
```

vs

```python
# Example 2
# %%
with BuildSketch() as sk:
    r1 = Rectangle(1,1)
    r2 = Rectangle(2,2)
    r2.locate(Pos(5,5))

show(sk.sketch)
```

I understand that BuildSketch() initialises on the default Plane.XY. I assume then, that each shape that is placed on this plane is centred on the origin (by default). So example2:r2 is centred on (0,0). Why cannot it then be relocated _on_ plane.XY to (5,5) using either `locate` or `move`?

**Comment by gumyr** [comment]
There are two important concepts as shown here:
```py
with BuildSketch() as sk:
    r1 = Rectangle(1, 1)
    with Locations((5, 5)):
        r2 = Rectangle(2, 2)

with BuildSketch(Pos(5, 5)) as sk2:
    r3 = Rectangle(1, 1)
    r4 = Rectangle(2, 2)
```
In `sk` a `r2` is being added at an arbitrary position on the default Plane.XY.  In `sk2`, the workplane that the sketch is using is not the default `Plane.XY` but a plane created from the `Location((5,5))` - i.e. Plane.XY but with the origin positioned at the global coordinate (5,5,0). As you mentioned, during construction the sketch is being built on Plane.XY but once it's complete it gets moved to one or more workplanes - in `sk2` that is the plane with an origin of (5,5).

In your "Example 2" `r2` is being moved after it has already been combined into the sketch so moving it doesn't do anything.

Outside of the builder, `sk.sketch` can be relocated with:
```py
sk.sketch.move(l) # relative move
sk.sketch.locate(l) # absolute move
sk.sketch.moved(l) # relative move with copy
sk.sketch.located(l) # absolute move with copy
sk.sketch.position = (x,y,z) # absolute move
sk.sketch.posiiton += (dx,dy,dz) # relative move
```

**Reply by weerasuriya** [reply]
Thanks.

So it seems that once the rectangle `r2` is instantiated _inside_ a sketch and combined into it, it is locked into the coordinate system of that parent sketch. To move `r2`, you move the whole parent sketch (and so if there are other shapes in that sketch, they would also move).

**Reply by gumyr** [reply]
> Thanks.
> 
> So it seems that once the rectangle `r2` is instantiated _inside_ a sketch and combined into it, it is locked into the coordinate system of that parent sketch. To move `r2`, you move the whole parent sketch (and so if there are other shapes in that sketch, they would also move).

As shown above, this is how to move `r2`:
```py
    with Locations((5, 5)):
        r2 = Rectangle(2, 2)
```

**Comment by jdegenstein** [comment]
Also, many of these concepts are covered in the introductory examples https://build123d.readthedocs.io/en/latest/introductory_examples.html I highly recommend you spend time working through these examples as it will save you time in the long run.

**Reply by weerasuriya** [reply]
Thanks. I have looked through the examples, tutorials, and API ref, but I could not find a specific explanation for this question (or if it exists, I missed it).
